lib/auth/iam_join.go

```
@@ -0,0 +1,301 @@
+
+func validateSTSIdentityRequest(req *http.Request, challenge string) error {
+	if req.Host != stsHost {
+		return trace.AccessDenied("sts identity request is for unknown host %q", req.Host)
+	}
+
+	if req.Method != http.MethodPost {
+		return trace.AccessDenied("sts identity request method %q does not match expected method %q", req.RequestURI, http.MethodPost)
+	}
+
+	if req.Header.Get(challengeHeaderKey) != challenge {
+		return trace.AccessDenied("sts identity request does not include challenge header or it does not match")
+	}
+
+	authHeader := req.Header.Get(authHeaderKey)
+	matches := signedHeadersRe.FindStringSubmatch(authHeader)
+	// first match should be the full header, second is the SignedHeaders
+	if len(matches) < 2 {
+		return trace.AccessDenied("sts identity request Authorization header is invalid")
+	}
+	signedHeadersString := matches[1]
+	signedHeaders := strings.Split(signedHeadersString, ";")
+	if !utils.SliceContainsStr(signedHeaders, normalizedChallengeHeaderKey) {
+		return trace.AccessDenied("sts identity request auth header %q does not include "+
+			normalizedChallengeHeaderKey+" as a signed header", authHeader)
+	}
+
+	// read and replace request body
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	req.Body = io.NopCloser(bytes.NewBuffer(body))
+
+	if !bytes.Equal([]byte(expectedSTSIdentityRequestBody), body) {
+		return trace.BadParameter("sts request body %q does not equal expected %q", string(body), expectedSTSIdentityRequestBody)
+	}
+
+	return nil
+}
+
+func parseSTSRequest(req []byte) (*http.Request, error) {
+	httpReq, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(req)))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Unset RequestURI and set req.URL instead (necessary quirk of sending a
+	// request parsed by http.ReadRequest). Also, force https here.
+	httpReq.RequestURI = ""
+	httpReq.URL = &url.URL{
```

Is there ever a path, etc, that might be silently dropped here?